### ISSUE TITLE
Node + Spawn a child process once and then provide stdin's and read stdout's from it

### ISSUE THREAD

**Issue by prashathsenthil** [nodejs/help#115]
I'm spawning a child process using spawn-command npm package, i do this when the node server starts and then for every request i read the query value and hit the running child process with stdin. The stdout that comes out of the child process is an event stream and i add a listener to read the data and then will be sending the data to sse-channel instance.

Everything works the way it should for the first request, the issue arises for subsequent requests. When i console.log the data inside the stdout's on listener handler then i could see that the data is getting cloned. Meaning, same set of data is getting repeated twice for the second request and thrice for the third request etc... I was researching this issue and possible solution for this would be to create independent streams for each request so that this data cloning would be avoided in the child.stdout's on listener handler. But i'm not sure whether this could resolve the issue. Could some one please suggest a way to overcome this hurdle.

Code looks like below,

```js 
var spawnCommand = require('spawn-command');
var cmd = 'path to the binary file'; 
 module.exports = function (app) {
  var child     = spawnCommand(cmd);
  srvc = {
   get: function(req, res) {
     child.stdin.write('a json object in a format that is expected by binary' + '\n');
     child.stdout.on('data', function() {
       console.log(''+ data);
     });
   }
  }
}
```

As i mentioned earlier the child process would be spawned once when the node server starts and then the get() method would run for every request. So each time when a request runs, the stdin's would be sent to the binary and then the "on" listener would log the data in stdout. This works perfectly for the first request, but when a second request is made (with the first request still logging data) i could see that the data's for both first and second are getting cloned twice and for third request the data's for each request is getting cloned thrice and it increases exponentially.

On the other hand if i spawn a child everytime inside the get() method i can overcome this problem but unfortunately i cannot do that because there'll be 100's of requests and i cannot afford 100's of binary instance eating up the memory.

Hope i explained it better, any suggestions would be of great help.

My noder version is 5.x.x (even though it does not matter with this query)

**Comment by lordKnighton** [comment]
How's it coming? I don't have much expertise with the **`child process` module**, have you tried adding an `'exit' or 'close'` on this code. 

so what if we add an `cmd.on('exit', function (exitCode) {
  console.log('exit', exitCode); `


- from https://github.com/mmalecki/spawn-command `README.md`
```js
var spawnCommand = require('spawn-command'),
    child = spawnCommand('echo "Hello spawn" | base64');

child.stdout.on('data', function (data) {
  console.log('data', data);
});

child.on('exit', function (exitCode) {
  console.log('exit', exitCode);
});
```

**Comment by prashathsenthil** [comment]
@lordKnighton i have an 'exit' listener which works when the node server is stopped (or) if the subscription request is cancelled. 

The stdout that streams out of the binary is a JSON  object and it would emit the data for every second. As i mentioned earlier if i'm going to spawn a new child process for every request i can very well avoid this data cloning issue but that would result in bad memory management since the binary would be spawned for every single request.  The 'pipes' array in socket object keeps growing like this when requests are made. 

     `pipes: [ [Object], [Object], [Object], [Object], [Object], [Object] ],
     pipesCount: 6 `

That might be the reason why the data is getting cloned, in this case there are 6 requests and in the console.log i could see  every data starting from first request is getting logged 6 times.

**Comment by prashathsenthil** [comment]
Still i'm stuck with this issue. Any luck from anyone?

**Comment by mh-cbon** [comment]
everytime a request hit your program, you ll attach a new handler to child.stoud. Assuming you calls `get` foreach request.

so what you could do is to off the listener foreach res.end event.

Something like this:
```
var spawnCommand = require('spawn-command');
var cmd = 'path to the binary file'; 
 module.exports = function (app) {
  var child     = spawnCommand(cmd);
  srvc = {
   get: function(req, res) {
     child.stdin.write('a json object in a format that is expected by binary' + '\n');
     var handle = function() {
       console.log(''+ data);
     };
     child.stdout.on('data', handle);
     res.on('finish', function () {
       child.stdout.removeEventListener(handle);
     })
   }
  }
}
```
wrote on the fly, to be carefully checked.

but i don t quiet understand how you are going to split the mixed stream data from the binary generated by two client requests.

**Comment by prashathsenthil** [comment]
@mh-cbon in this use case i'm dealing with event streams so if i understand correctly the `res.on('finish')` would run when the response is ended and then eventually as you mentioned it would remove the 'handle' listener, but since we are dealing with event streams here there can be a scenario where one or many request's could be made and then they'll be emitting the data's from binary continously unless that request is unsubscribed (for un-subscribe use case i have a different end point and that deals with that).. As far as mixed response is concerned each request has a unique UUID and that would be sent to the SSE channel... 

 Again the bottom line is, in this use case a request can be made and then the response would keep on streaming for dates/months so the chance of running this  `child.stdout.removeEventListener(handle);` can be rare... also when a request is active we can make another request and that's when i see data cloning issue. Having said that there is another issue which i foresee is Modern browsers restrict the number of simultaneous connections to the same origin to 6. If you attempt to make more connections at the same time, they will be blocked until active connections are closed (but this is different from what we are discussing here, thought of mentioning it).

**Comment by mh-cbon** [comment]
Sorry, what means SSE ?

> they'll be emitting the data from binary continously unless that request is unsubscribed 

Yes. It strongly depends of your program and what/how this binary works.

My remark was strictly about the fat that the binary has 1 stdout, on which you emits data for multiple clients, so the question is when you receive a data on the main and unique stdout, how do you re attribute it back to the correct client. Knowing that every connected client, will message from every other client.

A simplistic (? i m not very much experienced about this particular case) approach would be to split messages by line, and have a uuid at beginning of the line. so when you receive data, you can match clientUUID to lineUUID and determine if that message is about that client.

> So the bottom line in this use case is if a request is made, then the response would keep on streaming for dates/months hence the chance of running this child.stdout.removeEventListener(handle); can be rare... 

Once again, it strongly depends of your underlying bin. But yeah definitely possible if you don t implement an unsubscribe message between the response and the binary. Unsusbscribe `stdout.on('data')` event does not affect your underlying binary in any ways.

> Modern browsers restrict the number of simultaneous connections to the same origin to 6

Why would you open that much connections ? We usually do that to improve display speed of a web page by unblocking that limitation (which may be very ineffective because of the new dns resolutions it requires).
In your scenario i believe a websocket seems to match what you ve been describing so far, 
no ?

**Comment by codelahoma** [comment]
If by "binary" you mean an executable that you cannot modify, then you probably can't do what you're talking about.

The child process only has one `stdout` and one `stdin`.

The only way to do what you want would be to have a way of mapping events that are output by the binary back to the inputs that triggered them.

If this is the case, then all you should need to do is move the attachment of the `stdout`  handler out of the `get` function, so you only have one copy of it. Instead, attach it right after you spawn the child, at initialization.

**Comment by prashathsenthil** [comment]
@codelahoma thank you and that would be a feasible solution (but haven't tries that yet). On the other hand i tried something like below and with this i could eradicate the data cloning issue, fundamentally all i  did was checking the pipesCount in _readableState object which is originally a part of the socket response object and then resetting the pipes and _events array. With this i was able to avoid the cloning since i'm resetting the array each time when 'get' method runs.

Code would look something like this,

`if(child.stdout._readableState.pipesCount > 1) {`
            `child.stdout._readableState.pipes = new Array(child.stdout._readableState.pipes[0]);`
            `child.stdout._events.data = new Array(child.stdout._events.data[0]);`    
          `}`
